sdk/schemaregistry/schema-registry-avro/test/schemaRegistryAvroSerializer.spec.ts

```diff
@@ -0,0 +1,211 @@
+    await assert.isRejected(createTestSerializer().deserialize(Buffer.alloc(3)), /small/);
+  });
+
+  it("rejects invalid format", async () => {
+    const buffer = Buffer.alloc(42);
+    buffer.writeUInt32BE(0x1234, 0);
+    await assert.isRejected(createTestSerializer().deserialize(buffer), /format.*0x1234/);
+  });
+
+  it("rejects schema with no name", async () => {
+    const schema = JSON.stringify({ type: "record", fields: [] });
+    await assert.isRejected(createTestSerializer().serialize({}, schema), /name/);
+  });
+
+  it("rejects a schema with different serialization type", async () => {
+    const serializer = createTestSerializer();
+    const schema = await serializer.registry.registerSchema({
+      name: "_",
+      content: "_",
+      serializationType: "NotAvro",
+      group: "TestGroup"
+    });
+
+    const buffer = Buffer.alloc(36);
+    buffer.write(schema.id, 4, 32, "utf-8");
+    await assert.isRejected(
+      serializer.deserialize(buffer),
+      new RegExp(`${schema.id}.*NotAvro.*avro`)
+    );
+  });
+
+  it("serializes to the expected format", async () => {
+    const buffer = await createTestSerializer().serialize(testValue, testSchema);
+    assert.equal(0x0, buffer.readUInt32BE(0));
+    assert.equal(testSchemaIds[0], buffer.toString("utf-8", 4, 36));
+
+    const payload = buffer.slice(36);
+    assertSameJsonRepresentation(testAvroType.fromBuffer(payload), testValue);
+  });
+
+  it("deserializes from the expected format", async () => {
+    const serializer = createTestSerializer(false);
+    const payload = testAvroType.toBuffer(testValue);
+    const buffer = Buffer.alloc(36 + payload.length);
+
+    buffer.write(testSchemaIds[0], 4, 32, "utf-8");
+    payload.copy(buffer, 36);
+    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+  });
+
+  it("serializes and deserializes in round trip", async () => {
+    let serializer = createTestSerializer();
+    let buffer = await serializer.serialize(testValue, testSchema);
+    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+
+    // again for cache hit coverage on serialize
+    buffer = await serializer.serialize(testValue, testSchema);
+    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+
+    // throw away serializer for cache miss coverage on deserialize
+    serializer = createTestSerializer(false);
+    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+
+    // thow away serializer again and cover getSchemaId instead of registerSchema
+    serializer = createTestSerializer(false);
+    assert.deepStrictEqual(await serializer.serialize(testValue, testSchema), buffer);
+  });
+});
+
+function assertSameJsonRepresentation(actual: any, expected: any) {
+  // avsc returns objects with different prototype than our plain object literal
+  // so assert.deepEqual fails.
+  //
+  // REVIEW: Is it a problem that this avsc detail leaks?
```

Comment:
Can you explain more here? Wouldn't `assert.deepEqual` do what you want?

Reply:
assert.deepEqual did not do what I wanted but my explanation for why it did not might be wrong.  I'll put together a repro. I'm sure I'm going to learn something here. :)

Reply:
Still looking at this and will follow up separately. Added more info to #10950

Reply:
#11620 has the repro and tracks the issue now.